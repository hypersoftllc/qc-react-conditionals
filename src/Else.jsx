import React from 'react'


function Else(props) {
  let children = props.children,
      typeOf = typeof children

  return typeOf === 'number' || typeOf === 'string' ? children : children || null
}

export default Else
