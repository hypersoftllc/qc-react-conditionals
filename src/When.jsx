import PT from 'prop-types'
import React from 'react'


function When(props) {
  if (props.is) {
    let children, typeOf

    children = props.children
    typeOf = typeof children
    return typeOf === 'number' || typeOf === 'string' ? children : children || null
  }
  return null
}

When.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  is: PT.any,
}

export default When
