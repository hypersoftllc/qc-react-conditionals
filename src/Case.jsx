import React from 'react'

import Else from './Else'
import When from './When'


export default function Case(props) {
  let whenOrElse = false,
      children

  children = React.Children.map(props.children, (child) => {
    if (child.type === When) {
      if (!whenOrElse) {
        if (child.props.is) {
          whenOrElse = true
          return child
        }
      }
    }
    else if (child.type === Else) {
      if (!whenOrElse) {
        whenOrElse = true
        return child
      }
    }
    else {
      return child
    }
    return null
  })

  return children || null
}
