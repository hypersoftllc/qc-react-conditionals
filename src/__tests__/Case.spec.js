/* eslint-env jest, node */
/* eslint max-len: "off" */
/* eslint padded-blocks: "off" */

import React from 'react'
import TestRenderer from 'react-test-renderer'

import Case from '../Case'
import Else from '../Else'
import When from '../When'

const TRUE = true

describe('Case.jsx', function () {

  describe('Case', function () {

    it('should render with no props and no children', function () {
      let comp


      comp = TestRenderer.create(
        <Case/>
      )

      expect(comp.toJSON()).toMatchSnapshot()
    })

    describe('with no `When` or `Else` children', function () {

      it('should render the children as expected', function () {
        let comp


        comp = TestRenderer.create(
          <Case>
            First child.
            <span>Second child.</span>
            {
              [
                'Third child[0].',
                'Third child[1].',
              ]
            }
          </Case>
        )

        expect(comp.toJSON()).toMatchSnapshot()
      })

    })

    describe('with one `When` child', function () {

      describe('declared with a falsy `is` property', function () {

        it('should render the `When` component\'s children as `null`', function () {
          let comp


          comp = TestRenderer.create(
            <Case>
              <When>
                I am not rendered.
              </When>
            </Case>
          )

          expect(comp.toJSON()).toMatchSnapshot()


          comp = TestRenderer.create(
            <Case>
              <When is={false}>
                I am not rendered.
              </When>
            </Case>
          )

          expect(comp.toJSON()).toMatchSnapshot()
        })

      })

      describe('declared with a truthy `is` property', function () {

        it('should render the `When` component\'s children as expected', function () {
          let comp


          comp = TestRenderer.create(
            <Case>
              <When is={TRUE}>
                I am rendered.
              </When>
            </Case>
          )

          expect(comp.toJSON()).toMatchSnapshot()
        })

      })

      describe(', no `Else` children, and other non-`When` children', function () {

        describe('declared with a falsy `is` property', function () {

          it('should render the `When` component\'s children as `null` and the other non-`When` children as expected', function () {
            let comp


            comp = TestRenderer.create(
              <Case>
                I am always rendered.
                <When>
                  I am not rendered.
                </When>
                I am also rendered.
              </Case>
            )

            expect(comp.toJSON()).toMatchSnapshot()


            comp = TestRenderer.create(
              <Case>
                I am always rendered.
                <When>
                  I am not rendered.
                </When>
                I am also rendered.
              </Case>
            )

            expect(comp.toJSON()).toMatchSnapshot()
          })

        })

        describe('declared with a truthy `is` property', function () {

          it('should render the `When` component\'s children as expected and the other non-`When` children as expected', function () {
            let comp

            comp = TestRenderer.create(
              <Case>
                I am always rendered.
                <When is={TRUE}>
                  I am rendered.
                </When>
                I am also rendered.
              </Case>
            )

            expect(comp.toJSON()).toMatchSnapshot()
          })

        })

      })

      xdescribe('and one `Else` child', function () {

        it('', function () {

        })

      })

      xdescribe('and more than one `Else` child', function () {

        it('', function () {

        })

      })

    })

    describe('with more than one `When` child', function () {

      describe('where at least one `When` child with a truthy `is` property', function () {

        it('should render only the first truthy `When`\'s children as expected', function () {
          let comp


          // First truthy When is first When:
          comp = TestRenderer.create(
            <Case>
              <When is={TRUE}>
                I am rendered.
              </When>
              <When is={TRUE}>
                I am not rendered.
              </When>
              <When>
                I am not rendered.
              </When>
            </Case>
          )

          expect(comp.toJSON()).toMatchSnapshot()


          // First truthy When is intermediate When:
          comp = TestRenderer.create(
            <Case>
              <When is={false}>
                I am not rendered.
              </When>
              <When is={TRUE}>
                I am rendered.
              </When>
              <When is={TRUE}>
                I am not rendered.
              </When>
              <When>
                I am not rendered.
              </When>
            </Case>
          )

          expect(comp.toJSON()).toMatchSnapshot()


          // First truthy When is last When:
          comp = TestRenderer.create(
            <Case>
              <When is={false}>
                I am not rendered.
              </When>
              <When>
                I am not rendered.
              </When>
              <When is={TRUE}>
                I am rendered.
              </When>
            </Case>
          )

          expect(comp.toJSON()).toMatchSnapshot()
        })

      })

      xdescribe(', no `Else` children, and other non-`When` children', function () {

        it('', function () {

        })

      })

      xdescribe('and one `Else` child', function () {

        it('', function () {

        })

      })

      xdescribe('and more than one `Else` child', function () {

        it('', function () {

        })

      })

    })

    describe('with one or more `Else` child', function () {

      describe('and no `When` children', function () {

        it('should render the children as expected', function () {
          let comp


          comp = TestRenderer.create(
            <Case>
              First child.
              <span>Second child.</span>
              <Else>
                {
                  [
                    'Third child[0].',
                    'Third child[1].',
                  ]
                }
              </Else>
            </Case>
          )

          expect(comp.toJSON()).toMatchSnapshot()


          comp = TestRenderer.create(
            <Case>
              First child.
              <Else>
                <span>Second child.</span>
              </Else>
              {
                [
                  'Third child[0].',
                  'Third child[1].',
                ]
              }
            </Case>
          )

          expect(comp.toJSON()).toMatchSnapshot()


          comp = TestRenderer.create(
            <Case>
              <Else>
                First child.
              </Else>
              <span>Second child.</span>
              {
                [
                  'Third child[0].',
                  'Third child[1].',
                ]
              }
            </Case>
          )

          expect(comp.toJSON()).toMatchSnapshot()


          comp = TestRenderer.create(
            <Case>
              First child.
              <Else>
                <span>Second child.</span>
              </Else>
              <Else>
                {
                  [
                    'Third child[0].',
                    'Third child[1].',
                  ]
                }
              </Else>
            </Case>
          )

          expect(comp.toJSON()).toMatchSnapshot()


          comp = TestRenderer.create(
            <Case>
              <Else>
                First child.
              </Else>
              <span>Second child.</span>
              <Else>
                {
                  [
                    'Third child[0].',
                    'Third child[1].',
                  ]
                }
              </Else>
            </Case>
          )

          expect(comp.toJSON()).toMatchSnapshot()


          comp = TestRenderer.create(
            <Case>
              <Else>
                First child.
              </Else>
              <Else>
                <span>Second child.</span>
              </Else>
              {
                [
                  'Third child[0].',
                  'Third child[1].',
                ]
              }
            </Case>
          )

          expect(comp.toJSON()).toMatchSnapshot()
        })

      })

    })

  })

})
