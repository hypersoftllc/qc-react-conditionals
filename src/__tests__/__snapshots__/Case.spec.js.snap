// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Case.jsx Case should render with no props and no children 1`] = `null`;

exports[`Case.jsx Case with more than one \`When\` child where at least one \`When\` child with a truthy \`is\` property should render only the first truthy \`When\`'s children as expected 1`] = `"I am rendered."`;

exports[`Case.jsx Case with more than one \`When\` child where at least one \`When\` child with a truthy \`is\` property should render only the first truthy \`When\`'s children as expected 2`] = `"I am rendered."`;

exports[`Case.jsx Case with more than one \`When\` child where at least one \`When\` child with a truthy \`is\` property should render only the first truthy \`When\`'s children as expected 3`] = `"I am rendered."`;

exports[`Case.jsx Case with no \`When\` or \`Else\` children should render the children as expected 1`] = `
Array [
  "First child.",
  <span>
    Second child.
  </span>,
  "Third child[0].",
  "Third child[1].",
]
`;

exports[`Case.jsx Case with one \`When\` child , no \`Else\` children, and other non-\`When\` children declared with a falsy \`is\` property should render the \`When\` component's children as \`null\` and the other non-\`When\` children as expected 1`] = `
Array [
  "I am always rendered.",
  "I am also rendered.",
]
`;

exports[`Case.jsx Case with one \`When\` child , no \`Else\` children, and other non-\`When\` children declared with a falsy \`is\` property should render the \`When\` component's children as \`null\` and the other non-\`When\` children as expected 2`] = `
Array [
  "I am always rendered.",
  "I am also rendered.",
]
`;

exports[`Case.jsx Case with one \`When\` child , no \`Else\` children, and other non-\`When\` children declared with a truthy \`is\` property should render the \`When\` component's children as expected and the other non-\`When\` children as expected 1`] = `
Array [
  "I am always rendered.",
  "I am rendered.",
  "I am also rendered.",
]
`;

exports[`Case.jsx Case with one \`When\` child declared with a falsy \`is\` property should render the \`When\` component's children as \`null\` 1`] = `null`;

exports[`Case.jsx Case with one \`When\` child declared with a falsy \`is\` property should render the \`When\` component's children as \`null\` 2`] = `null`;

exports[`Case.jsx Case with one \`When\` child declared with a truthy \`is\` property should render the \`When\` component's children as expected 1`] = `"I am rendered."`;

exports[`Case.jsx Case with one or more \`Else\` child and no \`When\` children should render the children as expected 1`] = `
Array [
  "First child.",
  <span>
    Second child.
  </span>,
  "Third child[0].",
  "Third child[1].",
]
`;

exports[`Case.jsx Case with one or more \`Else\` child and no \`When\` children should render the children as expected 2`] = `
Array [
  "First child.",
  <span>
    Second child.
  </span>,
  "Third child[0].",
  "Third child[1].",
]
`;

exports[`Case.jsx Case with one or more \`Else\` child and no \`When\` children should render the children as expected 3`] = `
Array [
  "First child.",
  <span>
    Second child.
  </span>,
  "Third child[0].",
  "Third child[1].",
]
`;

exports[`Case.jsx Case with one or more \`Else\` child and no \`When\` children should render the children as expected 4`] = `
Array [
  "First child.",
  <span>
    Second child.
  </span>,
]
`;

exports[`Case.jsx Case with one or more \`Else\` child and no \`When\` children should render the children as expected 5`] = `
Array [
  "First child.",
  <span>
    Second child.
  </span>,
]
`;

exports[`Case.jsx Case with one or more \`Else\` child and no \`When\` children should render the children as expected 6`] = `
Array [
  "First child.",
  "Third child[0].",
  "Third child[1].",
]
`;
